File: app/src/main/java/com/giphy/browser/gif_list/GifListActivity.java
package com.giphy.browser.gif_list;

import android.content.Context;
import android.content.res.TypedArray;
import android.os.Bundle;
import android.view.Menu;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.SearchView;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;

import com.giphy.browser.GiphyApp;
import com.giphy.browser.R;
import com.giphy.browser.common.BaseActivity;
import com.giphy.browser.common.InfiniteScrollListener;
import com.giphy.browser.common.Repository;
import com.giphy.browser.databinding.ActivityGifListBinding;
import com.giphy.browser.gif_detail.GifDetailActivity;

import java.util.ArrayList;
import java.util.List;

public class GifListActivity extends BaseActivity {

    private ActivityGifListBinding binding;
    private GifListViewModel viewModel;
    private SearchView searchView;
    private GifListAdapter adapter;
    @Nullable
    private Toast toast;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_gif_list);

        final Context context = binding.getRoot().getContext();
        final TypedArray array = context.getResources().obtainTypedArray(R.array.backgroundColors);
        final List<Integer> backgroundColors = new ArrayList<>(array.length());
        for (int i = 0; i < array.length(); i++) {
            backgroundColors.add(array.getResourceId(i, 0));
        }
        array.recycle();

        final Repository repository = ((GiphyApp) getApplication()).getRepository();
        viewModel = new ViewModelProvider(this, new GifListViewModelFactory(repository, backgroundColors))
                .get(GifListViewModel.class);
        viewModel.getStateLiveData().observe(this, this::render);

        adapter = new GifListAdapter((position, item) -> viewModel.gifClicked(position, item));
        binding.content.setAdapter(adapter);
        final StaggeredGridLayoutManager layoutManager = new StaggeredGridLayoutManager(2, RecyclerView.VERTICAL);
        binding.content.setLayoutManager(layoutManager);
        binding.content.addOnScrollListener(new InfiniteScrollListener(
                layoutManager, 10, () -> viewModel.scrollThresholdReached()));

        binding.swipeContainer.setOnRefreshListener(() -> viewModel.screenRefreshed());
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.gif_list_menu, menu);
        searchView = (SearchView) menu.findItem(R.id.search).getActionView();

        searchView.setQuery(viewModel.getState().getQuery(), false);
        searchView.setIconified(!viewModel.getState().isSearchVisible());

        // Don't expand search view to full screen in landscape mode
        searchView.setImeOptions(searchView.getImeOptions() | EditorInfo.IME_FLAG_NO_EXTRACT_UI);
        searchView.setOnSearchClickListener((view) -> viewModel.searchClicked());
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                viewModel.queryUpdated(newText);
                return true;
            }
        });
        searchView.setOnCloseListener(() -> {
            viewModel.searchClosed();
            return false;
        });
        return true;
    }

    @Override
    public void onBackPressed() {
        if (!searchView.isIconified()) {
            searchView.setIconified(true);
        } else {
            super.onBackPressed();
        }
    }

    private void render(@NonNull GifListState state) {
        adapter.submitList(state.getItems());

        binding.progressBar.setVisibility(state.isLoading() ? View.VISIBLE : View.GONE);
        binding.swipeContainer.setRefreshing(state.isRefreshing());
        binding.swipeContainer.setEnabled(!state.isLoading());

        if (state.getToast() != null) {
            state.getToast().maybeConsume((strRes) -> {
                // Prevent multiple toasts from stacking
                if (toast != null) toast.cancel();
                toast = Toast.makeText(this, getString(strRes), Toast.LENGTH_SHORT);
                toast.show();
            });
        }

        if (state.getNavigateGifDetail() != null) {
            state.getNavigateGifDetail().maybeConsume((args) ->
                    startActivity(GifDetailActivity.newIntent(this, args)));
        }
    }
}

File: app/src/main/java/com/giphy/browser/gif_list/GifListActivity.kt
package com.giphy.browser.gif_list

import android.os.Bundle
import android.view.Menu
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.Toast
import androidx.appcompat.widget.SearchView
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.giphy.browser.GiphyApp
import com.giphy.browser.R
import com.giphy.browser.common.BaseActivity
import com.giphy.browser.common.InfiniteScrollListener
import com.giphy.browser.databinding.ActivityGifListBinding
import com.giphy.browser.gif_detail.GifDetailActivity

class GifListActivity : BaseActivity() {
    private lateinit var binding: ActivityGifListBinding
    private lateinit var viewModel: GifListViewModel
    private lateinit var searchView: SearchView
    private lateinit var adapter: GifListAdapter
    private var toast: Toast? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_gif_list)

        val array = resources.obtainTypedArray(R.array.backgroundColors)
        val backgroundColors = mutableListOf<Int>()
        for (i in 0 until array.length()) {
            backgroundColors.add(array.getResourceId(i, 0))
        }
        array.recycle()

        val repository = (application as GiphyApp).repository
        viewModel = createViewModel { GifListViewModel(repository, backgroundColors) }
        viewModel.stateLiveData
            .observe(this, Observer { render(it) })

        adapter = GifListAdapter(object : GifListAdapter.GifViewHolder.Listener {
            override fun onGifClicked(position: Int, item: GifItem) {
                viewModel.gifClicked(position, item)
            }
        })
        binding.content.adapter = adapter
        val layoutManager = StaggeredGridLayoutManager(2, RecyclerView.VERTICAL)
        binding.content.layoutManager = layoutManager
        binding.content.addOnScrollListener(
            InfiniteScrollListener(layoutManager, 10, object : InfiniteScrollListener.Listener {
                override fun onScrollThresholdReached() {
                    viewModel.scrollThresholdReached()
                }
            })
        )

        binding.swipeContainer.setOnRefreshListener { viewModel.screenRefreshed() }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.gif_list_menu, menu)
        searchView = menu!!.findItem(R.id.search).actionView as SearchView

        searchView.setQuery(viewModel.state.query, false)
        searchView.isIconified = !viewModel.state.isSearchVisible

        // Don't expand search view to full screen in landscape mode
        searchView.imeOptions = searchView.imeOptions or EditorInfo.IME_FLAG_NO_EXTRACT_UI
        searchView.setOnSearchClickListener { view: View? -> viewModel.searchClicked() }
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String) = true
            override fun onQueryTextChange(newText: String): Boolean {
                viewModel.queryUpdated(newText)
                return true
            }
        })
        searchView.setOnCloseListener {
            viewModel.searchClosed()
            false
        }
        return true
    }

    override fun onBackPressed() {
        if (!searchView.isIconified) {
            searchView.isIconified = true
        } else {
            super.onBackPressed()
        }
    }

    private fun render(state: GifListViewModel.State) {
        adapter.submitList(state.items)

        binding.progressBar.visibility = if (state.isLoading) View.VISIBLE else View.GONE
        binding.swipeContainer.isRefreshing = state.isRefreshing
        binding.swipeContainer.isEnabled = !state.isLoading

        state.toast?.maybeConsume {
            // Prevent multiple toasts from stacking
            toast?.cancel()
            toast = Toast.makeText(this, getString(it), Toast.LENGTH_SHORT)
            toast!!.show()
        }

        state.navigateGifDetail?.maybeConsume {
            startActivity(GifDetailActivity.newIntent(this, it))
        }
    }
}
